crates/binjs_io/src/escaped_wtf8.rs

```diff
@@ -0,0 +1,276 @@
+                let diff: u16 = (high as u16) * 64 + (low as u16);
+                let codepoint = LEAD_SORROGATE_MIN + diff;
+
+                buf.push(LONE_SORROGATE_ESCAPE_CHAR);
+                buf.push(encode_hex_char(((codepoint >> 12) & 0xf) as u8));
+                buf.push(encode_hex_char(((codepoint >> 8) & 0xf) as u8));
+                buf.push(encode_hex_char(((codepoint >> 4) & 0xf) as u8));
+                buf.push(encode_hex_char((codepoint & 0xf) as u8));
+            } else {
+                buf.push(input[end]);
+                buf.push(input[end + 1]);
+                buf.push(input[end + 2]);
+            }
+
+            begin = end + 3;
+        }
+
+        // ...... \xED \xA0 \x80 ......
+        //                       ^
+        //                       |
+        //                     begin
+
+        let pos = input[begin..].iter().position(|&c| c == LONE_SORROGATE_ESCAPE_CHAR);
+        if pos.is_none() {
+            buf.extend_from_slice(&input[begin..]);
+            break;
+        }
+
+        // ...... \xED \xA0 \x80 ...... \xED \xA0 \x80 ......
+        //                       ^      ^
+        //                       |      |
+        //                     begin   end
+        end = pos.unwrap();
+    };
+
+    buf
+}
+
+/// If the given `bytes` contains any escaped lone surropgate, unescape all
+/// escaped lone surrogate and returns the byte array.
+/// If not, return None.
+pub fn unescape(input: &[u8]) -> Option<Vec<u8>> {
+    let pos = input.iter().position(|&c| c == LONE_SORROGATE_ESCAPE_CHAR);
+    if pos.is_none() {
+        return None;
+    }
+
+    //   ...... \x7F XXXX ......
+    //   ^      ^
+    //   |      |
+    // begin   end
+    let mut begin = 0;
+    let mut end = pos.unwrap();
+
+    let mut buf: Vec<u8> = Vec::new();
+    loop {
+        let head = &input[begin..end];
+        buf.extend_from_slice(head);
+
+        let codepoint: u16 =
+            (decode_hex_char(input[end + 1]) << 12) |
+            (decode_hex_char(input[end + 2]) << 8) |
+            (decode_hex_char(input[end + 3]) << 4) |
+            decode_hex_char(input[end + 4]);
+        if codepoint == LONE_SORROGATE_ESCAPE_CHAR as u16 {
+            buf.push(LONE_SORROGATE_ESCAPE_CHAR);
+        } else {
+            assert!(codepoint >= LEAD_SORROGATE_MIN && codepoint <= TRAIL_SORROGATE_MAX,
+                    "escaped codepoint should be either escape character {:04x} or lone surrogate ({:04x}...{:04x})",
+                    LONE_SORROGATE_ESCAPE_CHAR,
+                    LEAD_SORROGATE_MIN,
+                    TRAIL_SORROGATE_MAX);
+
+            const LEAD_SURROGATE_MIN: u16 = 0xD800;
+            let diff = codepoint - LEAD_SURROGATE_MIN;
+            let high = (diff / 64) as u8;
+            let low = (diff % 64) as u8;
+            buf.push(LONE_SORROGATE_UNIT_1);
+            buf.push(LONE_SORROGATE_UNIT_2_MIN + high);
+            buf.push(LONE_SORROGATE_UNIT_3_MIN + low);
```

Comment:
Similar to in `escape`, I think this is correct but I half-expected the usual UTF-8 algorithm `0x80 | ((codepoint >> 6) & 0x3F)` and `0x80 | (codepoint & 0x3F)`